server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -1045,14 +1045,23 @@ public Symbol visitObjectLiteral(ObjectLiteral node, ExpressionAnalysisContext c
                 return Literal.EMPTY_OBJECT;
             }
             List<Symbol> arguments = new ArrayList<>(values.size() * 2);
+            var returnTypeBuilder = ObjectType.of(ColumnPolicy.DYNAMIC);
             for (Map.Entry<String, Expression> entry : values.entrySet()) {
+                var val = entry.getValue().accept(this, context);
                 arguments.add(Literal.of(entry.getKey()));
-                arguments.add(entry.getValue().accept(this, context));
+                arguments.add(val);
+                if (val.valueType() != UndefinedType.INSTANCE) {
+                    returnTypeBuilder.setInnerType(entry.getKey(), val.valueType());
+                }
```

Comment:
What's the reason to avoid setting the inner type if it is undefined? Wouldn't setting it anyway clarify that there is some value for the given key?

Reply:
Right, good spot, not needed and maybe even wrong (as you said). Unfortunately, I do not know anymore why this was needed, maybe because it wasn't needed at all ;)